### ISSUE TITLE
run in colab not pro

### ISSUE THREAD

**Issue by loboere** [rinongal/textual_inversion#4]
can you please create a notebook or someone has a notebook to train in colab not pro.

**Comment by rinongal** [comment]
I'll add this to my todos but getting it down to free colab level memory is going to be hard. If anyone wants to pick it up before me, feel free to do so and / or open a PR.

**Comment by GeorvityLabs** [comment]
@rinongal it's ok if it doesn't work in free tier, you can try.
But if you guys make a colab notebook for the pro level too it would be great , with the P100s.
Also , hoping that when you make the colab , users get info as to how many images they need to upload (used for finetuning)

**Comment by nicolai256** [comment]
> I'll add this to my todos but getting it down to free colab level memory is going to be hard. If anyone wants to pick it up before me, feel free to do so and / or open a PR.

i got this working on colab free tier and stable diffusion support 
https://colab.research.google.com/drive/1bnPJzT-dN0DUR50Z3GwtMtTKLm03o2c1?usp=sharing

**Comment by GeorvityLabs** [comment]
> > I'll add this to my todos but getting it down to free colab level memory is going to be hard. If anyone wants to pick it up before me, feel free to do so and / or open a PR.
> 
> i got this working on colab free tier and stable diffusion support https://colab.research.google.com/drive/1bnPJzT-dN0DUR50Z3GwtMtTKLm03o2c1?usp=sharing

few suggestions , in the notebook , the --init_word should be a one word description of whats in the image dataset right?
Also , because inside the .yaml file , the "*" is present , it will generate custom result for prompt : "a photo of *"

**Comment by GeorvityLabs** [comment]
so for example if you dataset is of faces , the shouldn't --init_word "person" ?

**Comment by GeorvityLabs** [comment]
any thoughts on the --init_word @rinongal

**Comment by nicolai256** [comment]
> 

> so for example if you dataset is of faces , the shouldn't --init_word "person" ?

yes it should be one word, if you do faces use --init_word "face"
i'm not 100% sure how it's working tbh, if it takes the `init_word` or only the `*` maybe rinongal knows :)

**Comment by rinongal** [comment]
The init_word and the placeholder token don't serve the same purpose. The placeholder token (which defaults to *) is the string that represents the new concept, so you leave it at *, you'll later use "a photo of *" to generate images of your concepts. If you change it to "@", you'll later need to use "a photo of @" to get the same effect etc.

The init_word is simply telling the model what to start the embedding of '*' from. Basically when the training starts, we grab the vector that represents the init_word, and give this same vector to '*'. From there, we optimize this vector to represent the specific concept. The init_word itself doesn't change at all and keeps the original vector.

If you're training on faces, I would use --init_word "face".

**Comment by nicolai256** [comment]
@rinongal 
i have 2 questions, is it possible to make the trainer save the checkpoints every 500 iterations?
and which file do you need to pass through the --resume command to resume the training?

**Comment by rinongal** [comment]
The trainer does save a checkpoint every roughly 500 iterations. Do you want to make it exactly 500? It's a little complicated but it can be done.

Resume is a actually one of LDM's original arguments. It's not updated for our work. I'll add resume support to the to-do, but it's very simple so I'll try to get to it early tomorrow.

**Comment by nicolai256** [comment]
> The trainer does save a checkpoint every roughly 500 iterations. Do you want to make it exactly 500? It's a little complicated but it can be done.
> 
> Resume is a actually one of LDM's original arguments. It's not updated for our work. I'll add resume support to the to-do, but it's very simple so I'll try to get to it early tomorrow.

yeah basically save a .pt and .ckpt file every 500 iterations, would be good to test which is the best one to use
and thanks!
i've also added a thing to save the images with an extra number at the end when trying to resume(because there was a saving issue for images when doing --resume), it's in my version of the repo
https://github.com/nicolai256/Stable-textual-inversion_win
(sorry for not forking it, i downloaded, changed and uploaded it but i gave u credit :) )

**Comment by rinongal** [comment]
If all you want to do is check previous checkpoints - you can already do that. We save an embedding_gs-××××.pt file every roughly 500 iterations. It's in the log directory under checkpoints.

It's just roughly every 500 checkpoints and not exactly every 500 because it was easier to build that on top of the existing pytorch lightning code and I was lazy :) Do you have a use case that requires a checkpoint exactly every 500 iterations?

**Comment by nicolai256** [comment]
> If all you want to do is check previous checkpoints - you can already do that. We save an embedding_gs-××××.pt file every roughly 500 iterations. It's in the log directory under checkpoints.
> 
> It's just roughly every 500 checkpoints and not exactly every 500 because it was easier to build that on top of the existing pytorch lightning code and I was lazy :) Do you have a use case that requires a checkpoint exactly every 500 iterations?

oh i might have borked something in the code then, now it only saves when stopping the training
i'll check what fucked it up

**Comment by rickyars** [comment]
@nicolai256 thanks for the notebook! any idea on how to get the inference to work? e.g., in the notebook i'm getting the same error as you:
``` 
Traceback (most recent call last):
  File "scripts/txt2img_text_inv.py", line 10, in <module>
    from ldm.util import instantiate_from_config
ModuleNotFoundError: No module named 'ldm'
```

**Comment by rickyars** [comment]
> 

fwiw, this is my experience as well. i have to kill the run for the checkpoint to save.

**Comment by GeorvityLabs** [comment]
> > The trainer does save a checkpoint every roughly 500 iterations. Do you want to make it exactly 500? It's a little complicated but it can be done.
> > Resume is a actually one of LDM's original arguments. It's not updated for our work. I'll add resume support to the to-do, but it's very simple so I'll try to get to it early tomorrow.
> 
> yeah basically save a .pt and .ckpt file every 500 iterations, would be good to test which is the best one to use (it does save one every epoch now tho so maybe it's possible to fix it through that) and thanks! i've also added a thing to save the images with an extra number at the end when trying to resume(because there was a saving issue for images when doing --resume), it's in my version of the repo https://github.com/nicolai256/Stable-textual-inversion_win (sorry for not forking it, i downloaded, changed and uploaded it but i gave u credit :) )

also , when trying to generate , i get the following issue.
any ideas on how to fix it.
![Uploading Screenshot from 2022-08-25 08-48-05.png…]()

**Comment by GeorvityLabs** [comment]
> > If all you want to do is check previous checkpoints - you can already do that. We save an embedding_gs-××××.pt file every roughly 500 iterations. It's in the log directory under checkpoints.
> > It's just roughly every 500 checkpoints and not exactly every 500 because it was easier to build that on top of the existing pytorch lightning code and I was lazy :) Do you have a use case that requires a checkpoint exactly every 500 iterations?
> 
> oh i might have borked something in the code then, now it only saves when stopping the training i'll check what fucked it up

![Screenshot from 2022-08-25 08-48-05](https://user-images.githubusercontent.com/58555432/186566845-57b2ce5d-f1c2-4b79-8ca7-761b001163bb.png)
also , when trying to generate , i get the following issue.
any ideas on how to fix it.
@nicolai256 @rinongal

**Comment by nicolai256** [comment]
> > > If all you want to do is check previous checkpoints - you can already do that. We save an embedding_gs-××××.pt file every roughly 500 iterations. It's in the log directory under checkpoints.
> > > It's just roughly every 500 checkpoints and not exactly every 500 because it was easier to build that on top of the existing pytorch lightning code and I was lazy :) Do you have a use case that requires a checkpoint exactly every 500 iterations?
> > 
> > 
> > oh i might have borked something in the code then, now it only saves when stopping the training i'll check what fucked it up
> 
> ![Screenshot from 2022-08-25 08-48-05](https://user-images.githubusercontent.com/58555432/186566845-57b2ce5d-f1c2-4b79-8ca7-761b001163bb.png) also , when trying to generate , i get the following issue. any ideas on how to fix it. @nicolai256 @rinongal

that text to image script is not working sorry, i'll clean it up

**Comment by GeorvityLabs** [comment]
> > > > If all you want to do is check previous checkpoints - you can already do that. We save an embedding_gs-××××.pt file every roughly 500 iterations. It's in the log directory under checkpoints.
> > > > It's just roughly every 500 checkpoints and not exactly every 500 because it was easier to build that on top of the existing pytorch lightning code and I was lazy :) Do you have a use case that requires a checkpoint exactly every 500 iterations?
> > > 
> > > 
> > > oh i might have borked something in the code then, now it only saves when stopping the training i'll check what fucked it up
> > 
> > 
> > ![Screenshot from 2022-08-25 08-48-05](https://user-images.githubusercontent.com/58555432/186566845-57b2ce5d-f1c2-4b79-8ca7-761b001163bb.png) also , when trying to generate , i get the following issue. any ideas on how to fix it. @nicolai256 @rinongal
> 
> that text to image script is not working sorry, i'll clean it up

cool , would be great to verify the finetuned model.
I have downloaded the checkpoints, will wait for the fix.

**Comment by GeorvityLabs** [comment]
also @rinongal , how many checkpoints do you think it will take to accurately recreate a face , given 5 images of the faces in the training dataset.
What would be your recommended xxxx number for the embedding_gs-××××.pt file. (3000 , 5000 ?)
Were you able to run any tests.

**Comment by GeorvityLabs** [comment]
@rinongal , I just had a look at the images generated @ 10,000 iterations and inside the validation set, looks pretty good.

**Comment by rinongal** [comment]
We never tried faces, so you're welcome to experiment and let us know :) 
The images in the paper all used models with 5k iterations.

**Comment by rickyars** [comment]
> 

i was able to get it to work and the results are pretty awesome. can you show me your full prompt? i had to drop the samples and iters to avoid out-of-memory issues but it seems to be working. 

also, any idea why the checkpoint file is so small? i was expecting GB file size but it's only kB.

**Comment by rickyars** [comment]
> 

i ran overnight and got 43k iterations. seems like overkill and not sure i like the results. will be nice to get those 500 iteration saves implemented. can't be of much help this week, but perhaps next week i'll have some free time to dig into the code.

**Comment by rinongal** [comment]
The checkpoint files contain only learned embeddings. 

We're not modifying the model, only learning a code that represents a new word. These codes are roughly 1000 numbers (a bit more if you use LDM, a bit less if you use SD), so the file containing them is just a few KB.

**Comment by rickyars** [comment]
> We're not modifying the model, only learning a code that represents a new word. These codes are rough

great - this is what i assumed was happening but nice to have the confirmation

**Comment by GeorvityLabs** [comment]
> > 
> 
> i was able to get it to work and the results are pretty awesome. can you show me your full prompt? i had to drop the samples and iters to avoid out-of-memory issues but it seems to be working.
> 
> also, any idea why the checkpoint file is so small? i was expecting GB file size but it's only kB.

were you able to run custom prompts? 
I was able to train and get the checkpoint files.
But, SD keeps crashing when trying to run "a photo of *" , @nicolai256 said he was working on a fix.

**Comment by nicolai256** [comment]
> 

i have a fix for local but not for colab, i dont really use colab tbh..
my modified repo recently got merged into https://github.com/lstein/stable-diffusion
so u could use that, it has prompt weighting img2img text inversion, all of the good stuff and it's getting updated every couple hours

**Comment by rickyars** [comment]
> > > 
> > 
> > 
> > i was able to get it to work and the results are pretty awesome. can you show me your full prompt? i had to drop the samples and iters to avoid out-of-memory issues but it seems to be working.
> > also, any idea why the checkpoint file is so small? i was expecting GB file size but it's only kB.
> 
> were you able to run custom prompts? I was able to train and get the checkpoint files. But, SD keeps crashing when trying to run "a photo of *" , @nicolai256 said he was working on a fix.

yeah - it's working great for me. e.g., here is "death metal cover featuring *
![0016](https://user-images.githubusercontent.com/3086631/186702993-e743720c-0baa-471f-9e11-3182371d4ae9.jpg)
"

**Comment by loboere** [comment]
@nicolai256  @rinongal  amazing job but how can i resume a checkpoint in colab?

**Comment by GeorvityLabs** [comment]
> > > > 
> > > 
> > > 
> > > i was able to get it to work and the results are pretty awesome. can you show me your full prompt? i had to drop the samples and iters to avoid out-of-memory issues but it seems to be working.
> > > also, any idea why the checkpoint file is so small? i was expecting GB file size but it's only kB.
> > 
> > 
> > were you able to run custom prompts? I was able to train and get the checkpoint files. But, SD keeps crashing when trying to run "a photo of *" , @nicolai256 said he was working on a fix.
> 
> yeah - it's working great for me. e.g., here is "death metal cover featuring *"
> 
> ![0016](https://user-images.githubusercontent.com/3086631/186702993-e743720c-0baa-471f-9e11-3182371d4ae9.jpg)

@rickyars do you have the notebook/script that you used to generate the results from the checkpoint , would be great if you have a notebook , i could test it on my end.

**Comment by rinongal** [comment]
@loboere Resume from checkpoint will be added over the weekend. I don't control the colab so you'll need to wait for whoever maintains it to add it there.

**Comment by nicolai256** [comment]
> @loboere Resume from checkpoint will be added over the weekend. I don't control the colab so you'll need to wait for whoever maintains it to add it there.

i control the colab, i can give you writing permissions if you want or i can add it when you add it :)

**Comment by GeorvityLabs** [comment]
> > > > 
> > > 
> > > 
> > > i was able to get it to work and the results are pretty awesome. can you show me your full prompt? i had to drop the samples and iters to avoid out-of-memory issues but it seems to be working.
> > > also, any idea why the checkpoint file is so small? i was expecting GB file size but it's only kB.
> > 
> > 
> > were you able to run custom prompts? I was able to train and get the checkpoint files. But, SD keeps crashing when trying to run "a photo of *" , @nicolai256 said he was working on a fix.
> 
> i have a fix for local but not for colab, i dont really use colab tbh.. my modified repo recently got merged into https://github.com/lstein/stable-diffusion so u could use that, it has prompt weighting img2img text inversion, all of the good stuff and it's getting updated every couple hours

do you have the fix in the form of a jupyter notebook (.ipynb) , then I could test it out on my local gpu.
Would be great if you could upload it on drive , I can download it and test it out locally on my GPU.
Currently , I have the downloaded checkpoint (.pt files) , would be great to see what script you run in order to generate from the checkpoint.

**Comment by nicolai256** [comment]
> 

you'll have to ask [lstein](https://github.com/lstein/stable-diffusion) about that

i do have the stable_txt2img.py and img2img.py modified in my fork and they work with textual inversion but lsteins repo is much easier and better

**Comment by nicolai256** [comment]
> If all you want to do is check previous checkpoints - you can already do that. We save an embedding_gs-××××.pt file every roughly 500 iterations. It's in the log directory under checkpoints.
> 
> It's just roughly every 500 checkpoints and not exactly every 500 because it was easier to build that on top of the existing pytorch lightning code and I was lazy :) Do you have a use case that requires a checkpoint exactly every 500 iterations?

figured out how to let it save every 500 steps :)

add this to lightning module in v1-finetune.yaml

lightning:
  modelcheckpoint:
    params:
      every_n_train_steps: 500
  callbacks:

**Comment by rinongal** [comment]
The config file now has every_n_train_steps: 500 on by default (thanks @nicolai256)

To resume training from a given checkpoint you can add `--embedding_manager_ckpt <path to existing embeddings file>` to your command.

**Comment by rinongal** [comment]
Closing due to lack of activity. Feel free to reopen if you still need help.

**Comment by HannesGitH** [comment]
> > I'll add this to my todos but getting it down to free colab level memory is going to be hard. If anyone wants to pick it up before me, feel free to do so and / or open a PR.
> 
> i got this working on colab free tier with stable diffusion support https://colab.research.google.com/drive/1bnPJzT-dN0DUR50Z3GwtMtTKLm03o2c1?usp=sharing

it currently fails with `RuntimeError: indices should be either on cpu or on the same device as the indexed tensor (cpu)` in my colab free

sd-v1-4.ckpt is the correct model, right?
